libgalois/src/analytics/jaccard/jaccard.cpp

```diff
@@ -0,0 +1,250 @@
+    // Collect all the neighbors of the base node into a hash set.
+    for (const auto& e : graph.edges(base)) {
+      auto dest = graph.GetEdgeDest(e);
+      base_neighbors.emplace(*dest);
+    }
+  }
+
+  uint32_t operator()(GNode n2) {
+    uint32_t intersection_size = 0;
+    for (const auto& e : graph_.edges(n2)) {
+      auto neighbor = graph_.GetEdgeDest(e);
+      if (base_neighbors.count(*neighbor) > 0)
+        intersection_size++;
+    }
+    return intersection_size;
+  }
+};
+
+template <typename IntersectAlgorithm>
+galois::Result<void>
+JaccardImpl(
+    galois::graphs::PropertyGraph<std::tuple<JaccardSimilarity>, std::tuple<>>&
+        graph,
+    size_t compare_node, JaccardPlan /*plan*/) {
+  if (compare_node >= graph.size()) {
+    return galois::ErrorCode::InvalidArgument;
+  }
+
+  auto it = graph.begin();
+  std::advance(it, compare_node);
+  Graph::Node base = *it;
+
+  uint32_t base_size = graph.edge_end(base) - graph.edge_begin(base);
+
+  IntersectAlgorithm intersect_with_base{graph, base};
+
+  // Compute the similarity for each node
+  galois::do_all(
+      galois::iterate(graph),
+      [&](const GNode& n2) {
+        double& n2_data = graph.GetData<JaccardSimilarity>(n2);
+        uint32_t n2_size = graph.edge_end(n2) - graph.edge_begin(n2);
+        // Count the number of neighbors of n2 and the number that are shared
+        // with base
+        uint32_t intersection_size = intersect_with_base(n2);
+        // Compute the similarity
+        uint32_t union_size = base_size + n2_size - intersection_size;
+        double similarity =
+            union_size > 0 ? (double)intersection_size / union_size : 1;
+        // Store the similarity back into the graph.
+        n2_data = similarity;
+      },
+      galois::loopname("Jaccard"));
+
+  return galois::ResultSuccess();
+}
+
+}  // namespace
+
+galois::Result<void>
+galois::analytics::Jaccard(
+    graphs::PropertyFileGraph* pfg, size_t compare_node,
+    const std::string& output_property_name, JaccardPlan plan) {
+  if (auto result =
+          ConstructNodeProperties<NodeData>(pfg, {output_property_name});
+      !result) {
+    return result.error();
+  }
+
+  auto pg_result = Graph::Make(pfg, {output_property_name}, {});
+  if (!pg_result) {
+    return pg_result.error();
+  }
+
+  galois::Result<void> r = galois::ResultSuccess();
+  switch (plan.edge_sorting()) {
+  case JaccardPlan::kUnknown:
+    // TODO(amp): It would be possible to start with the sorted case and then
+    //  fail to the unsorted case if unsorted nodes are detected.
+  case JaccardPlan::kUnsorted:
+    r = JaccardImpl<IntersectWithUnsortedEdgeList>(
+        pg_result.value(), compare_node, plan);
+    break;
+  case JaccardPlan::kSorted:
+    r = JaccardImpl<IntersectWithSortedEdgeList>(
+        pg_result.value(), compare_node, plan);
+    break;
+  }
+
+  if (!r) {
+    // Undo property creation.
+    if (auto r1 = pfg->RemoveNodeProperty(output_property_name); !r1) {
```

Comment:
Do we need this? If we do not mark properties as persistent that should suffice, isn't it?

Reply:
Ah. Yes. I meant to ask about this.

I think we do need it. The reason is that leaving the property on the `PropertyFileGraph` will: a) force a rerun of the algorithm to use a different property name to avoid conflicts, b) waste memory since the old property is referenced and will hang around in memory. These are both issue in a Jupyter notebook where having a failure in a given cell is usually handled by only rerunning that one cell.

We could fix this by changing the property creation semantics of algorithms to be "add or overwrite", but I think that could make it easy to accidentally destroy your data due to the silent overwrite.

Reply:
I'll remove it for now and create an issue to track this and fix it eventually. One way or the other.

Reply:
oh ok. Makes sense to remove it if it produces invalid results. I was thinking of just doing a single routine on a graph but one can do multiple routines and these properties will continue to grow.